### ISSUE TITLE
operator to get a shuffled static method ref

### ISSUE THREAD

**Issue by gavinking** [ceylon/ceylon-spec#1370]
I would really like to be able to write this:

```ceylon
{ "hello", "goodbye" }.map(String!plus(" world"))
```

Or perhaps even:

```ceylon
{ "hello", "goodbye" }.map(String.plus!(" world"))
```

Here `!` is basically just a shortcut for `shuffle()`, though it could be compiled to something more optimal.

Yesyes we can already write the above using `Iterable.spread()`:

```ceylon
{ "hello", "goodbye" }.spread(String.plus)(" world")
```

but that really doesn't handle all usecases. What about this case, for example:

```ceylon
{ "hello", "goodbye" }.every(String!longerThan(10))
```

**Comment by gavinking** [comment]
Indeed, you could even generalize this operator to regular functions, so that:

    (1..20).map(formatInteger!(16))

Would mean:

    (1..20).map(shuffle(curry(formatInteger))(16))

i.e. that is:

    (1..20).map((i) => formatInteger(i, 16))

**Comment by FroMage** [comment]
What does `shuffle` do? Its name implies random card shuffling and I can't imagine this is what it does to parameters.

**Comment by FroMage** [comment]
Sounds like what you want to do is partial application of the `n+1` parameters. What about partial application of the first `n` parameters? Or the first and last `n,m` parameters? Where do we draw the line?

**Comment by gavinking** [comment]
<https://modules.ceylon-lang.org/repo/1/ceylon/language/1.1.0/module-doc/api/index.html#shuffle>

**Comment by FroMage** [comment]
We should seriously consider embargoing every new function to the language module so we can agree on its name. This one is seriously misleading. But then, so are `flatten` and `unflatten` already.

**Comment by gavinking** [comment]
@FroMage it is quite common to want to apply a function to a list of the thing with the same type as its *first* parameter. It's *especially* common when the function is a method, and the first "parameter" is the receiver.

**Comment by gavinking** [comment]
@FroMage Every _new_ function? That one has been there since like before there was a typechecker.

**Comment by FroMage** [comment]
I agree with you in principle, but I've also seen that other languages with partial application tend to favour the first `n` parameters, which is at odds with this particular case. And there _must_ be a reason for them to support this, so I suppose it also must be common, no?

**Comment by FroMage** [comment]
Well, it was new when it was added, and I never saw it ;)

**Comment by gavinking** [comment]
It is every bit as venerable as `compose()`, `curry()` and `uncurry()`. It predates `flatten()` and `unflatten()` IIRC.

**Comment by FroMage** [comment]
And look, I'm not saying I would have named it better, but you have to admit its name is almost contrary to what it does: it's not random at all.

**Comment by FroMage** [comment]
`compose` and `curry` are well-established FP terms. `flatten` is a disastrous name, but I've never found better.

**Comment by jvasileff** [comment]
A couple thoughts:

* I like the idea, it *seems* like it would be very useful
* I don't like the `!`, because it 1) looks strange, and 2) is what others use for an `assert exists`/unsafe unwrap operator

Would the argument affect type inference? Normally, with the shuffle approach, we have:

```ceylon
    print(["abcd"].map((s) => s.map(Character.integer)));
    print(["abcd"].map(shuffle(String.map<Integer>)(Character.integer)));
```
where `Integer` is not inferred. This may seem a bit convoluted, but it matches the immediate case I found where this might be useful:

```ceylon
return that.declarations.flatMap(
        //(d) => d.transform(topLevelTransformer)
        shuffle(Node.transform<[DartCompilationUnitMember*]>)(topLevelTransformer));
```

**Comment by sirinath** [comment]
This is partial application

```
(1..20).map(@papply(1) formatInteger(16))
```

```
(1..20).map(@papply(1..2) formatInteger(16, 32))
```
